### PR TITLE
[connector/datadog] fix lifecycle

### LINKED ISSUE #30487: [connector/datadog] Shutdown hangs
### Component(s)

_No response_

### What happened?

When running collector components, we need to make sure that we can shutdown without starting them first. This is the expectation set forth with https://github.com/open-telemetry/opentelemetry-collector/issues/6507.

When running the datadog connector with traces, we see that shutdown hangs.

I have created a test to show how to reproduce this issue:
```
func Test_Shutdown_Traces(t *testing.T) {
	factory := NewFactory()

	creationParams := connectortest.NewNopCreateSettings()
	cfg := factory.CreateDefaultConfig().(*Config)

	traceToMetricsConnector, err := factory.CreateTracesToMetrics(context.Background(), creationParams, cfg, consumertest.NewNop())
	require.NoError(t, err)
	err = traceToMetricsConnector.Shutdown(context.Background())
	require.NoError(t, err)
}
```

See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30486 for a PR with the test.

The fix consists in changing the behavior of the connector to check whether the exitChan is active, by moving its initialization to the Start function.

### Collector version

latest

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
fix start logic to start if tracesConsumer is present; make shutdown work even if start was not called

**Related issue**
Fixes #30487

### CODE REVIEW

**Comment by atoulme** [comment]
Depends on #30471 to be merged first.

**Review by songy23 — APPROVED** [review]
cc @dineshg13

File: connector/datadogconnector/connector.go [hunk 1]
```diff
@@ -67,15 +67,15 @@ func newConnector(set component.TelemetrySettings, _ component.Config, metricsCo
...
 	}, nil
 }
 
 // Start implements the component.Component interface.
 func (c *connectorImp) Start(_ context.Context, _ component.Host) error {
 	c.logger.Info("Starting datadogconnector")
 	c.agent.Start()
-	if c.metricsConsumer != nil {
+	c.exit = make(chan struct{})
+	if c.metricsConsumer != nil || c.tracesConsumer != nil {
```
Comment (songy23) [inline connector/datadogconnector/connector.go:78]: It is intentional that we only do `go c.run` for only `metricsConsumer`. We don't want to start it in traces pipelines. Please revert.

File: connector/datadogconnector/connector.go [hunk 2]
```diff
@@ -67,15 +67,15 @@ func newConnector(set component.TelemetrySettings, _ component.Config, metricsCo
...
-		exit:            make(chan struct{}),
 	}, nil
 }
 
 // Start implements the component.Component interface.
 func (c *connectorImp) Start(_ context.Context, _ component.Host) error {
 	c.logger.Info("Starting datadogconnector")
 	c.agent.Start()
-	if c.metricsConsumer != nil {
+	c.exit = make(chan struct{})
```
Comment (songy23) [inline connector/datadogconnector/connector.go:77]: Why this change? Seems no change compared to doing it in `newConnector`

**Review by songy23 — CHANGES_REQUESTED** [review]
Sorry pressed the approve button by mistake

File: connector/datadogconnector/connector.go (hunk 1 above)
Comment (atoulme) [reply connector/datadogconnector/connector.go:78]: alright - I'll open a separate PR with a test reproducing the problem and file an issue.

Comment (atoulme) [reply connector/datadogconnector/connector.go:78]: #30487 is filed

File: connector/datadogconnector/connector.go (hunk 2 above)
Comment (atoulme) [reply connector/datadogconnector/connector.go]: See #30487 for reasoning

**Review by songy23 — COMMENTED** [review]
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31026 should unblock this.

**Comment by songy23** [comment]
Please rebase

**Comment by atoulme** [comment]
@songy23 rebased

**Review by songy23 — COMMENTED** [review]
`Generated code is out of date, please run "make generate" and commit the changes in this PR.`

Needs another rebase too

**Comment by dmitryax** [comment]
@atoulme can you please rebase?

**Comment by atoulme** [comment]
Done, will ping once tests complete if they pass.

**Comment by atoulme** [comment]
Tests didn't pass. It looks like the "trace to metrics" path still fails. @songy23 what do you think?

**Review by songy23 — COMMENTED** [review]
Hmm I was under the impression that the recent changes should have fixed the shutdown, but it seems a different issue.

We will take a closer look and fix it to unblock this PR. cc @mackjmr

**Comment by mx-psi** [comment]
@mackjmr can you take a look?

**Comment by mackjmr** [comment]
@atoulme put up #31811 to fix this, I think we can close this PR in favor of new one.

**Comment by atoulme** [comment]
Closing in favor of #31811
